File: app/src/main/java/com/kimjisub/launchpad/network/UnishareDownloader.java
package com.kimjisub.launchpad.network;

import android.os.AsyncTask;

import com.kimjisub.launchpad.api.file.FileApi;
import com.kimjisub.launchpad.api.unipad.vo.UnishareVO;
import com.kimjisub.launchpad.manager.Unipack;
import com.kimjisub.manager.FileManager;
import com.kimjisub.manager.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;

import okhttp3.ResponseBody;
import retrofit2.Call;

public class UnishareDownloader {
	UnishareVO unishare;
	File F_UniPackRootExt;
	UnishareDownloadListener listener;

	File zip;
	File folder;

	public interface UnishareDownloadListener {
		void onDownloadStart();

		void onGetFileSize(long fileSize);

		void onDownloading(long downloadedSize);

		void onDownloadEnd(File zip);

		void onAnalyzeStart();

		void onAnalyzeSuccess(Unipack unipack);

		void onAnalyzeFail(Unipack unipack);

		void onAnalyzeEnd(File folder);

		void onException(Throwable throwable);
	}

	public UnishareDownloader(UnishareVO unishare, File F_UniPackRootExt, UnishareDownloadListener listener) {
		this.unishare = unishare;
		this.F_UniPackRootExt = F_UniPackRootExt;
		this.listener = listener;

		zip = FileManager.makeNextPath(F_UniPackRootExt, unishare.title + " #" + unishare._id, ".zip");
		folder = FileManager.makeNextPath(F_UniPackRootExt, unishare.title + " #" + unishare._id, "/");

		(new DownloadThread()).execute();
	}

	class DownloadThread extends AsyncTask<String, String, String> {

		@Override
		protected String doInBackground(String... strings) {
			try {
				listener.onDownloadStart();
				Call<ResponseBody> call = FileApi.getService().unishare_download(unishare._id);
				ResponseBody responseBody = call.execute().body();
				InputStream in = responseBody.byteStream();

				listener.onGetFileSize(responseBody.contentLength());

				FileOutputStream out = new FileOutputStream(zip);
				byte[] buf = new byte[1024];
				long downloadSize = 0L;
				int n;
				int loop = 0;
				while (-1 != (n = in.read(buf))) {
					out.write(buf, 0, n);
					downloadSize += n != -1 ? n : 0;
					if (loop % 100 == 0)
						listener.onDownloading(downloadSize);
					loop++;
				}
				in.close();
				out.close();

				listener.onDownloadEnd(zip);


				listener.onAnalyzeStart();

				FileManager.unZipFile(zip.getPath(), folder.getPath());
				Unipack unipack = new Unipack(folder, true);
				if (unipack.CriticalError) {
					Log.err(unipack.ErrorDetail);

					listener.onAnalyzeFail(unipack);
					FileManager.deleteDirectory(folder);
				} else
					listener.onAnalyzeSuccess(unipack);

				listener.onAnalyzeEnd(folder);

			} catch (Exception e) {
				e.printStackTrace();
				listener.onException(e);
				FileManager.deleteDirectory(folder);
			}
			FileManager.deleteDirectory(zip);

			return null;
		}
	}
}

File: app/src/main/java/com/kimjisub/launchpad/network/UnishareDownloader.kt
package com.kimjisub.launchpad.network

import com.kimjisub.launchpad.api.file.FileApi.Companion.service
import com.kimjisub.launchpad.api.unipad.vo.UnishareVO
import com.kimjisub.launchpad.manager.Unipack
import com.kimjisub.manager.FileManager
import com.kimjisub.manager.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.io.OutputStream

class UnishareDownloader(
		private val unishare: UnishareVO,
		UniPackRootExt: File,
		private var listener: UnishareDownloadListener) {
	private val zip: File = FileManager.makeNextPath(UniPackRootExt, unishare.title + " #" + unishare._id, ".zip")
	private val folder: File = FileManager.makeNextPath(UniPackRootExt, unishare.title + " #" + unishare._id, "/")

	interface UnishareDownloadListener {
		fun onDownloadStart()
		fun onGetFileSize(contentSize: Long, realSize: Long)
		fun onDownloading(downloadedSize: Long)
		fun onDownloadEnd(zip: File)
		fun onAnalyzeStart()
		fun onAnalyzeSuccess(unipack: Unipack)
		fun onAnalyzeFail(unipack: Unipack)
		fun onAnalyzeEnd(folder: File)
		fun onException(throwable: Throwable)
	}

	init {
		CoroutineScope(Dispatchers.Main).launch {
			try {
				listener.onDownloadStart()

				var inputStream: InputStream? = null
				var outputStream: OutputStream?

				var fileSize: Long? = null

				withContext(Dispatchers.IO) {
					val call = service.unishare_download(unishare._id)
					val responseBody = call!!.execute().body()
					inputStream = responseBody!!.byteStream()
					fileSize = responseBody.contentLength()
				}
				listener.onGetFileSize(fileSize!!, if (fileSize!! > 0) fileSize!! else unishare.fileSize)

				withContext(Dispatchers.IO) {
					outputStream = FileOutputStream(zip)
					val buf = ByteArray(1024)
					var n: Int
					var downloadSize = 0L
					var loop = 0
					while (true) {
						n = inputStream!!.read(buf)
						if (n == -1)
							break

						outputStream!!.write(buf, 0, n)
						downloadSize += n.toLong()
						if (loop % 100 == 0)
							withContext(Dispatchers.Main) { listener.onDownloading(downloadSize) }
						loop++
					}
					inputStream!!.close()
					outputStream!!.close()
				}

				listener.onDownloadEnd(zip)
				listener.onAnalyzeStart()

				var unipack: Unipack? = null
				withContext(Dispatchers.IO) {
					FileManager.unZipFile(zip.path, folder.path)
					unipack = Unipack(folder, true)
				}
				if (unipack!!.CriticalError) {
					Log.err(unipack!!.ErrorDetail)
					listener.onAnalyzeFail(unipack!!)
					withContext(Dispatchers.IO) {
						FileManager.deleteDirectory(folder)
					}
				} else
					listener.onAnalyzeSuccess(unipack!!)
				listener.onAnalyzeEnd(folder)
			} catch (e: Exception) {
				e.printStackTrace()
				listener.onException(e)
				withContext(Dispatchers.IO) {
					FileManager.deleteDirectory(folder)
				}
			}
			withContext(Dispatchers.IO) {
				FileManager.deleteDirectory(zip)
			}
		}
	}
}